wallet/src/de/schildbach/wallet/ui/dashpay/CreateIdentityService.kt

```diff
@@ -0,0 +1,228 @@
+
+    override fun onCreate() {
+        super.onCreate()
+        try {
+            securityGuard = SecurityGuard()
+        } catch (e: Exception) {
+            log.error("Unable to instantiate SecurityGuard", e)
+            stopSelf()
+            return
+        }
+        createIdentityNotification.startServiceForeground()
+    }
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        super.onStartCommand(intent, flags, startId)
+
+        if (intent != null) {
+
+            when (intent.action) {
+                ACTION_CREATE_IDENTITY -> handleCreateIdentityAction(intent)
+            }
+        }
+
+        return START_NOT_STICKY
+    }
+
+    private fun handleCreateIdentityAction(intent: Intent) {
+        val username = intent.getStringExtra(EXTRA_USERNAME)
+
+        val exceptionHandler = CoroutineExceptionHandler { _, exception ->
+            log.error("[${identityCreationState.state}(error)]", exception)
+            GlobalScope.launch {
+                updateState(identityCreationState.state, true)
+//                stopSelf()
+            }
+        }
+
+        serviceScope.launch(exceptionHandler) {
+            createIdentity(username)
+            stopSelf()
+        }
+    }
+
+    private suspend fun createIdentity(username: String) {
+
+        //identityCreationState = identityCreationStateDaoAsync.load()
+        //        ?: IdentityCreationState(IdentityCreationState.State.UPGRADING_WALLET, false, username)
+        identityCreationState = IdentityCreationState(IdentityCreationState.State.UPGRADING_WALLET, false, username)
+        identityCreationStateDaoAsync.insert(identityCreationState)
+
+        if (identityCreationState.state != IdentityCreationState.State.UPGRADING_WALLET || identityCreationState.error) {
+            log.info("resuming identity creation process [${identityCreationState.state}${if (identityCreationState.error) "(error)" else ""}]")
+        }
+
+        val handler = Handler()
+        val wallet = walletApplication.wallet
+        val password = securityGuard.retrievePassword()
+
+        val encryptionKey = deriveKey(handler, wallet, password)
+        val seed = decryptSeed(handler, wallet, encryptionKey)
+
+//        val usernameInfo = CreateUsernameInfo(username, seed, encryptionKey)
+
+
+        platformRepo.addWalletAuthenticationKeysAsync(seed, encryptionKey)
+
+        updateState(IdentityCreationState.State.CREDIT_FUNDING_TX_SENDING)
+
+        //create the Blockchain Identity object (this needs to be saved somewhere eventually)
+        val blockchainIdentity = BlockchainIdentity(Identity.IdentityType.USER, 0, wallet)
+
+        platformRepo.createCreditFundingTransactionAsync(blockchainIdentity, encryptionKey)
+
+        walletApplication.broadcastTransaction(blockchainIdentity.creditFundingTransaction)
+
+        updateState(IdentityCreationState.State.CREDIT_FUNDING_TX_SENT)
+
```

Comment:
Is the code below for testing purposes only?

Reply:
Some of it is for testing purposes.

The 2 second delay between steps is for testing purposes (to watch process bars).

Currently there are several states for the Credit Funding Transaction (Create, Sending, Sent, Confirmed), but not all may be necessary to keep track of.  Also, this is still missing the code to determine that the transaction was sent and acknowledged by the network.  

The remaining steps that handle registering the identity and username are working.

Reply:
Delays are removed: https://github.com/dashevo/dash-wallet/pull/393/commits/df09a470bc2902a7daa055eb72faffd22280083b